### PR TITLE
make placement respect symbol-sort-key

### LINKED ISSUE #8896: `symbol-sort-key` placement behavior appears to fail along tile boundaries when `icon-allow-overlap` is `false`.
**mapbox-gl-js version**: 1.4.1

**browser**: Chrome 77.0.3865.120 and Safari 13.0.1 (14608.2.11.1.11)

### Steps to Trigger Behavior

 1. Add a symbol layer with a source that contains Feature 1 with a lower `symbol-sort-key` and a Feature 2 with a higher `symbol-sort-key` whose icon collides with the Feature 1 but is in the tile either to the west or north of the tile of the Feature 1.

### Link to Demonstration

The demonstration allows you to toggle between Feature 2 being in the tile to the west, east, north, or south of the tile of Feature 1. Green icon is the Feature 1, red icon is the Feature 2. 
https://jsbin.com/favowijosu/2/edit?js,output

### Expected Behavior
Feature 1 (having the lower `symbol-sort-key`) should be shown and Feature 2 (having the higher `symbol-sort-key`) should be hidden due to collision.

### Actual Behavior
Feature 2 is shown and Feature 1 is hidden.

The expected behavior _is_ seen if Feature 2 is in the tile to the east or south of Feature 1.

### PR DESCRIPTION
**Pull request by ansis**
Currently symbol-sort-key works:
- across tile boundaries with -allow-overlap: true
- within tiles for collision detection

This pr makes it work for collisions across tile boundaries as well.

The implementation
- creates a BucketPart for each range of a bucket that has the same sort key
- sorts all the buckets across tile boundaries if needed
- does placement in the correct order

fix #8896

## Launch Checklist

 - [x] briefly describe the changes in this PR
 - [see render tests] include before/after visuals or gifs if this PR includes visual changes
 - [x] write tests for all new functionality
 - [n/a] document any changes to public APIs
 - [x] post benchmark scores (**This still needs to be done carefully**)
 - [x] manually test the debug page
 - [n/a] tagged `@mapbox/studio` and/or `@mapbox/maps-design` if this PR includes style spec or visual changes
 - [ ] tagged `@mapbox/gl-native` if this PR includes shader changes or needs a native port
 - [x] apply changelog label ('bug', 'feature', 'docs', etc) or use the label 'skip changelog'
 - [x] `<changelog>Fix bug where `symbol-sort-key` was not used for collisions that crossed tile boundaries. </changelog>`

### CODE REVIEW

**Comment by ansis** [comment]
I added a placement benchmark and it looks like this keeps performance close enough to what it was before.

Screen Shot 2019-12-05 at 5 48 32 PM

**Comment by ansis** [comment]
This approach could be extended a bit further to allow symbols to be sorted globally for collisions. This would let you have different symbol layers but intermingle the collision priority.

There are some style-spec design challenges associated with that:
- we would need a new property or a change to the existing `symbol-sort-key` that enables global sort order or provides a separate global sort key
- this small implementation change would only impact collisions, not rendering for symbols that allow overlap. Currently symbol-sort-key affects both.

This change would be out of the scope of this pr.

**Comment by ahk** [comment]
> Hi, sorry but I think there may be a bug. I get some unstable behavior on tile boundaries.

I think I had something weird with my dev server, that behavior is now gone and the feature works correctly. Sorry about the runaround.

File: src/style/pauseable_placement.js [hunk 1]
```diff
@@ -6,28 +6,56 @@ import {Placement} from '../symbol/placement';
...
+    _sortAcrossTiles: boolean;
     _currentTileIndex: number;
-    _tiles: Array<Tile>;
+    _currentPartIndex: number;
     _seenCrossTileIDs: { [string | number]: boolean };
+    _bucketParts: Array<BucketPart>;
+
+    constructor(styleLayer: SymbolStyleLayer) {
+        this._sortAcrossTiles = styleLayer.type === 'symbol' &&
+            styleLayer.layout.get('symbol-z-order') !== 'viewport-y' &&
```
[the remark this answers is not in the document]
Comment (ansis) [reply src/style/pauseable_placement.js]: documenting conversation elsewhere: This matches existing behavior. The existing behavior and documentation don't match clearly. We need to update either the docs or change the behavior.

The unclear case is when `symbol-z-order` === `source` and there is a sort key set.

File: src/symbol/placement.js [hunk 2]
```diff
@@ -289,15 +327,30 @@ export class Placement {
         }
     }
 
-    placeLayerBucket(bucket: SymbolBucket, posMatrix: mat4, textLabelPlaneMatrix: mat4, iconLabelPlaneMatrix: mat4,
-            scale: number, textPixelRatio: number, showCollisionBoxes: boolean, holdingForFade: boolean, seenCrossTileIDs: { [string | number]: boolean },
-            collisionBoxArray: ?CollisionBoxArray) {
-        const layout = bucket.layers[0].layout;
-        const partiallyEvaluatedTextSize = symbolSize.evaluateSizeForZoom(bucket.textSizeData, this.transform.zoom);
+    placeLayerBucket(bucketPart: Object, seenCrossTileIDs: { [string | number]: boolean }, showCollisionBoxes: boolean) {
```
[the remark this answers is not in the document]
Comment (ansis) [reply src/symbol/placement.js]: fixed in 398e5f9

File: src/style/pauseable_placement.js [hunk 3]
```diff
@@ -6,28 +6,56 @@ import {Placement} from '../symbol/placement';
...
 class LayerPlacement {
+    _sortAcrossTiles: boolean;
     _currentTileIndex: number;
-    _tiles: Array<Tile>;
+    _currentPartIndex: number;
     _seenCrossTileIDs: { [string | number]: boolean };
+    _bucketParts: Array<BucketPart>;
+
+    constructor(styleLayer: SymbolStyleLayer) {
+        this._sortAcrossTiles = styleLayer.type === 'symbol' &&
```
[the remark this answers is not in the document]
Comment (ansis) [reply src/style/pauseable_placement.js]: fixed in 398e5f9

File: src/style/pauseable_placement.js [hunk 4]
```diff
@@ -6,28 +6,56 @@ import {Placement} from '../symbol/placement';
...
+            styleLayer.layout.get('symbol-sort-key').constantOr(1) !== undefined;
 
-    constructor() {
         this._currentTileIndex = 0;
+        this._currentPartIndex = 0;
         this._seenCrossTileIDs = {};
+        this._bucketParts = [];
     }
 
     continuePlacement(tiles: Array<Tile>, placement: Placement, showCollisionBoxes: boolean, styleLayer: StyleLayer, shouldPausePlacement: () => boolean) {
```
[the remark this answers is not in the document]
Comment (ansis) [reply src/style/pauseable_placement.js]: fixed in 398e5f9

**Pull request merged**